auth_pysaml/models/res_users.py

```diff
@@ -0,0 +1,209 @@
+
+    def _auth_saml_signin(self, provider, validation, saml_response):
+        """ retrieve and sign into openerp the user corresponding to provider
+        and validated access token
+
+            :param provider: saml provider id (int)
+            :param validation: result of validation of access token (dict)
+            :param saml_response: saml parameters response from the IDP
+            :return: user login (str)
+            :raise: openerp.exceptions.AccessDenied if signin failed
+
+            This method can be overridden to add alternative signin methods.
+        """
+        token_osv = self.env["auth_saml.token"]
+        saml_uid = validation["user_id"]
+        user = (
+            self.env["res.users.saml"]
+            .search(
+                [("saml_uid", "=", saml_uid), ("saml_provider_id", "=", provider)],
+                limit=1,
+            )
+            .user_id
+        )
+        if len(user) != 1:
+            raise AccessDenied()
+        # now find if a token for this user/provider already exists
+        token_ids = token_osv.search(
+            [("saml_provider_id", "=", provider), ("user_id", "=", user.id)]
+        )
+
+        if token_ids:
+            token_ids.write({"saml_access_token": saml_response})
+        else:
+            _logger.info("Creating auth_saml.token")
+            token_ids = token_osv.create(
+                {
+                    "saml_access_token": saml_response,
+                    "saml_provider_id": provider,
+                    "user_id": user.id,
+                }
+            )
+
+        if validation.get("mapped_attrs", {}):
+            user.write(validation.get("mapped_attrs", {}))
+
+        return user.login
+
+    @api.model
+    def auth_saml(self, provider, saml_response, base_url=None):
+        validation = self._auth_saml_validate(provider, saml_response, base_url)
+
+        # required check
+        if not validation.get("user_id"):
+            raise AccessDenied()
+
+        # retrieve and sign in user
+        login = self._auth_saml_signin(provider, validation, saml_response)
+
+        if not login:
+            raise AccessDenied()
+
+        # return user credentials
+        return self.env.cr.dbname, login, saml_response
+
+    def _check_credentials(self, token):
+        """Override to handle SAML auths.
+
+        The token can be a password if the user has used the normal form...
+        but we are more interested in the case when they are tokens
+        and the interesting code is inside the "except" clause.
+        """
+        try:
+            # Attempt a regular login (via other auth addons) first.
+            super(ResUser, self)._check_credentials(token)
+
+        except (AccessDenied, passlib.exc.PasswordSizeError):
+            # since normal auth did not succeed we now try to find if the user
+            # has an active token attached to his uid
+            res = (
```

Comment:
Res will be the returned result

```suggestion
            user = (
```